source/http2_stream_manager.c

```diff
@@ -0,0 +1,801 @@
+            STREAM_MANAGER_LOGF(
+                DEBUG,
+                stream_manager,
+                "connection %p reaches max concurrent streams limits. Streams assigned to the connection=%zu, "
+                "while limits=%" PRIu32 ". Moving it out of available connections.",
+                (void *)chosen_connection->connection,
+                chosen_connection->num_streams_assigned,
+                chosen_connection->max_concurrent_streams);
+        }
+        AWS_ASSERT(re_code == AWS_ERROR_SUCCESS && "random access set went wrong");
+    }
+    return;
+}
+
+/**
+ * It can be invoked from:
+ * - User release last refcount of stream manager
+ * - User acquires stream from stream manager
+ * - Connection acquired callback from connection manager
+ * - Stream compeleted callback from HTTP
+ */
+/* *_synced should only be called with LOCK HELD or from another synced function */
+static void s_aws_http2_stream_manager_build_transaction_synced(struct aws_http2_stream_management_transaction *work) {
+    struct aws_http2_stream_manager *stream_manager = work->stream_manager;
+    if (stream_manager->synced_data.state == AWS_H2SMST_READY) {
+
+        /* Steps 1: Pending acquisitions of stream */
+        while (!aws_linked_list_empty(&stream_manager->synced_data.pending_acquisitions)) {
+            struct aws_linked_list_node *node =
+                aws_linked_list_pop_back(&stream_manager->synced_data.pending_acquisitions);
+            struct aws_h2_sm_pending_stream_acquisition *pending_acquisition =
+                AWS_CONTAINER_OF(node, struct aws_h2_sm_pending_stream_acquisition, node);
+            s_get_sm_connection_synced(stream_manager, pending_acquisition);
+            if (pending_acquisition->sm_connection == NULL) {
+                /* Cannot find any connection, push it back and break the loop */
+                aws_linked_list_push_back(&stream_manager->synced_data.pending_acquisitions, node);
+                STREAM_MANAGER_LOGF(
+                    TRACE,
+                    stream_manager,
+                    "acquisition:%p cannot find any connection to use.",
+                    (void *)pending_acquisition);
+                break;
+            } else {
+                /* found connection for the request. Move it to pending make requests and update the count */
+                aws_linked_list_push_back(&work->pending_make_requests, node);
+                --stream_manager->synced_data.pending_acquisition_count; /* Length of the pending_acquisitions */
+                ++stream_manager->synced_data.scheduled_stream_count;
+            }
+        }
+
+        /* Step 2: Check for new connections needed */
+        if (stream_manager->synced_data.pending_acquisition_count) {
+            size_t num_connections_needed = stream_manager->synced_data.pending_acquisition_count / UINT32_MAX + 1;
```

Comment:
this math is weird
you're telling me it should just be 1
and then I say: then this can just be a bool


we talked about how complicated this will eventaully be, so just do the simplest possible thing now

Reply:
changed to be compare with the ideal